Sources/ArgumentParser/Parsing/ArgumentDefinition.swift

```diff
@@ -39,8 +39,13 @@ struct ArgumentDefinition {
   
   struct Help {
     var options: Options
-    var help: ArgumentHelp?
-    var discussion: String?
+
+    // `ArgumentHelp` members
+    @NonEmpty var abstract: String?
+    @NonEmpty var discussion: String?
+    @NonEmpty var valueName: String?
```

Comment:
Is there a semantic difference for these between `""` and `nil`? If not, I'd rather make these non-optional and just use empty strings.

Reply:
I don't believe so, just that the property is derived from `help: ArgumentHelp?` so the optionality is inherited from there. I am happy to make these non-optional and make the points of use test for empty string. I think the only rough point is losing the ergonomics of optional chaining and defaulting. ex: `valueName ?? otherValueName ?? "default"`

Reply:
Example: <img width="1358" alt="Screen Shot 2021-05-20 at 9 03 32 AM" src="https://user-images.githubusercontent.com/9739930/119034638-4453e780-b94a-11eb-9cb1-a4748573680b.png">

Reply:
Example:
```diff
-    return help.valueName
-      ?? preferredNameForSynopsis?.valueString
+    return help.valueName.isEmpty
+      ? help.valueName
+      : preferredNameForSynopsis?.valueString
       ?? help.keys.first?.rawValue.convertedToSnakeCase(separator: "-")
       ?? "value"
```

Reply:
I think this is less desirable as it makes it easier to accidentally introduce bugs, like I just did in the above code. The condition should be `*!*help.valueName.isEmpty`

Reply:
Yeah, the ergonomics are annoying, but I'd rather not keep them optional just for that purpose. That's kind of a funny usage point, since it's then lifting `help.valueName` into an optional to use with the following optional chaining… You could add a `String` extension like `func mapEmpty(_ f: () -> String) -> String` that would clear that up:

```swift
help.valueName.mapEmpty {
    preferredNameForSynopsis?.valueString
       ?? help.keys.first?.rawValue.convertedToSnakeCase(separator: "-")
       ?? "value"
}
```

Reply:
I went with you suggestion, but as an extension on `Collection` instead, since this replacement behavior seems generally useful.

---

It almost seems like there should be a `NullValue` protocol, such that this extension could operate on anything with a defined `null` and then the `??` operator could be extended to as well.

```swift
protocol NullValue {
  static var null: Self
}

extension NullValue where Self: Equatable {
  func mapNull(_ replacement: () -> Self) -> Self {
    self == null ? replacement() : self
  }
}

func ??<T: NullValue>(lhs: T, rhs: @autoclosure T) -> T {
  lhs.mapNull(rhs)
}
```

Reply:
Yeah, it’s been proposed, I think!